File: Simulator.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
	char buf[120];
	strcpy(buf, argv[1]);
	strcat(buf, ".mac");
	
	FILE *P;
	P = fopen(buf, "rb");
	
	//Registers
	short ac = 0, pc = 0, sp = 0, ir;
	short opcode, address;
	
	//Main memory
	short mem[4096];
	
	fread(mem, 2, 4096, P);
	
	while(1){
		
		//1. Fetch
		ir = mem[pc];
		
		//get the opcode
		opcode = ir & 0XF000;
		opcode = opcode >> 12;
		
		//get the address
		address = ir & 0X0FFF;
		
		//2. Increment pc
		pc++;
		
		//3. Decode and 4. Execute
		switch(opcode)
		{
			//load instruction
			case 0X0000: ac = mem[address];
					break;
			
			//store instruction
			case 0X0001: mem[address] = ac;
					break;
			
			//add instruction
			case 0X0002: ac = ac + mem[address];
					break;
			
			//sub instruction
			case 0X0003: ac = ac - mem[address];
					break;
			
			//push instruction
			case 0X0004: mem[--sp] = ac;
					break;
			
			//aloc instruction
			case 0X0005: sp = sp - address;
					break;
			
			//dloc instruction
			case 0X0006: sp = sp + address;
					break;
			
			//din instruction
			case 0X0007: scanf("%hd",&ac);
					break;
			
			//ldc instruction
			case 0X0008: ac = address;
			
			//ja instruction
			case 0X0009: pc = address; break;
			
			//ret instruction
			case 0X000A: pc = mem[sp++]; break;
			
			//sout instruction
			case 0X000B: while(mem[ac] != 0)
						printf("%c",mem[ac++]);
					break;
			
			//dout instruction
			case 0X000C: printf("%d",ac);
					break;
			
			//jnz instruction
			case 0X000D: if(ac != 0)
					pc = address;
					break;
			
			//call instruction
			case 0X000E: mem[--sp] = pc;
					pc = address;
					break;
			
			//halt instruction
			case 0X000F: exit(0);
		}
	}
}

File: Simulator.java
import java.nio.file.*;
import java.io.*;
import java.nio.*;
import java.util.*;

public class Simulator {

  public static void main(String[] args) {

  try {

  File file = new File("C:/Users/Akshay/Downloads/mystest.mac");
  byte[] arrayOfBytes = new byte[(int) file.length()];
  FileInputStream fis = new FileInputStream(file);
  fis.read(arrayOfBytes);
  fis.close();

  /*System.out.println("memory size is:"+arrayOfBytes.length);
  for(int i=0; i<arrayOfBytes.length; i++)
	System.out.println("first element of memory is:"+arrayOfBytes[i]);*/

  short[] memory = new short[arrayOfBytes.length/2];

  ByteBuffer.wrap(arrayOfBytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(memory);
  /*System.out.println("memory size is:"+memory.length);
  for(int i=0; i<memory.length; i++)
	System.out.println("first element of memory is:"+memory[i]);*/



  short pc=0,sp=0,ac=0,ir;              // Creating registers

  while(true){
    ir = memory[pc];
    pc++;                               // Increment PC register by 1
    //System.out.printf("%hd\n", ir);
    int opCode,address;

    opCode = ir & 0XF000;               // Seperate opCode from instruction
    address = ir & 0X0FFF;              // Seperate address from instruction

    opCode = opCode >> 12;              // Shifting opCode to left
    //System.out.printf("%d  %d\n", opCode, address);
    switch(opCode){

      case 0X0000 :                     // Load instruction
        ac = memory[address];
        break;

      case 0X0001 :                     // Store instruction
        memory[address] = ac;
        break;

      case 0X0002 :                     // Add instruction
        ac += memory[address];
        break;

      case 0X0003 :                     // Sub instruction
        ac -= memory[address];
        break;

      case 0X0004 :                     // Push instruction
        if (sp == 0)
          sp = (short)(arrayOfBytes.length/2);
        memory[--sp] = ac;
        break;

      case 0X0005 :                     // Aloc instruction
        if(sp == 0)
          sp = (short)(arrayOfBytes.length/2);
        sp -= address;
        break;

      case 0X0006 :                     // Dloc instruction
        sp += address;
        if(sp >= (short)(arrayOfBytes.length/2))
          sp -= (short)(arrayOfBytes.length/2);
        break;

      case 0X0007 :                     // Din instruction
        Scanner sc = new Scanner(System.in);
        ac = sc.nextShort();
        break;

      case 0x0008 :                     // Ldc instruction
        ac = (short)address;
        break;

      case 0X0009 :                     // Ja instruction
        pc = (short)address;
        break;

      case 0X000A :                     // Ret instruction
        pc = memory[sp++];
        break;

      case 0X000B :                     // Sout instruction
        while(memory[ac] != 0)
        System.out.printf("%c",memory[ac++]);
        break;

      case 0X000C :                     // Dout instruction
        System.out.printf("%d", ac);
        break;

      case 0X000D :                     // Jnz instruction
        if(ac != 0)
          pc = (short)address;
        break;

      case 0X000E :                     // Call instruction
        memory[--sp] = pc;
        pc = (short)address;
        break;

      case 0X000F :                     // Halt instruction
        System.exit(0);

    }
  }
    } catch (IOException e) {
              e.printStackTrace();
          }
}
}
